impl/src/main/java/io/jsonwebtoken/impl/lang/DefaultCollectionMutator.java

```diff
@@ -48,7 +49,45 @@ private boolean doAdd(E e) {
 
     @Override
     public M add(E e) {
```

Comment:
This is a fairly complex implementation - it almost makes me think we should create a specific abstract subclass for Algorithm/ID-specific logic, e.g. `AlgorithmCollectionMutator` or something like that.

Such a subclass could also be more efficient, potentially using a `Set<String>` or a `Set<String,Identifiable>` to keep track of what IDs exist already instead of iterating over the internal collection.

Finally, there's no method contract that collection element ordering will remain unchanged, so we don't need to perform step 3 - that will implicitly create a performance penalty by saving trailing elements and adding them back in to the end of the collection after replacing the found element.

Thoughts?